File: scripts/nemertea/nemertea.py
import time
import random

app: "AppProxy" # type: ignore

class State:
    NONE = 0
    TESTING = 1
    ACTIVE = 2
    INVALID = 3
    
# NODE --------------------------------------------------------------------------

class Node:
    def __init__(self, vertex):
        self.vertex = vertex
        self.parent = None
        self.edge = None
        self.children = []

    def add_child(self, vertex):
        child = Node(vertex)
        child.parent = self
        child.edge = self.vertex.get_edge_to(vertex)
        self.children.append(child)
        return child

    def destroy(self):
        if self.parent and self.edge and self.edge.get_state() != State.ACTIVE:
            self.edge.set_state(State.NONE)

        if self.vertex.get_state() != State.ACTIVE:
            self.vertex.set_state(State.NONE)

        self.parent = None
        self.children = None

# NBFS --------------------------------------------------------------------------

class NBFS:

    def __init__(self, vertex, first):
        self.root = Node(vertex)
        self.first = first
        self.level = 0
        self.leaves = [self.root]
        self.found = False
        self.node = None
        self.new_vertices = 0

    def run(self):
        root = self.root
        first = self.first

        while self.leaves and not self.found and self.level < app.get_var("deep"):
            new_level = []
            for leaf in self.leaves:
                app.step()
                edge_size = leaf.vertex.get_edge_size()

                for i in range(edge_size):
                    app.step()
                    edge = leaf.vertex.get_edge(i)
                    child, self.found = self.select_child(root, leaf, edge, first)

                    if self.found:
                        self.node = child
                        break
                    elif child:
                        new_level.append(child)

                if app.is_stopped() or self.found:
                    break

            if not self.found:
                self.level += 1
                self.leaves = new_level

            if app.is_stopped():
                return

        if self.found:
            self.make_path(self.node)

        self.destroy(self.root)
        return self.new_vertices

    def select_child(self, root, node, edge, first):
        # Caso 1: Aresta já ativa - ignora
        if edge.get_state() == State.ACTIVE:
            return None, False

        adjacent = node.vertex.get_adjacent(edge)

        # Caso 2: Só na primeira iteração pode visitar vértices TESTING
        if not first and adjacent.get_state() == State.TESTING:
            return None, False

        # Caso 3: Ignora voltar para o pai
        if node.parent and adjacent.get_id() == node.parent.vertex.get_id():
            return None, False

        # Caso 4: Se é raiz sem conexões ativas ainda
        if adjacent.get_id() == root.vertex.get_id() and root.vertex.get_active_edge_size() == 0:
            return node.add_child(adjacent), True

        # Caso 5 e 6: Se vértice está ativo e é vizinho da raiz, caminho encontrado ou ignora
        if adjacent.get_state() == State.ACTIVE:
            edge_to_root = adjacent.get_edge_to(root.vertex)
            if edge_to_root and edge_to_root.get_state() == State.ACTIVE:
                return node.add_child(adjacent), True
            return None, False

        # Caso 7: Vértice livre, adiciona como filho
        edge.set_state(State.TESTING)
        adjacent.set_state(State.TESTING)
        return node.add_child(adjacent), False

    def make_path(self, node):
        # Se existir conexão ativa com raiz, desconecta para abrir caminho novo
        if node.vertex.get_state() == State.ACTIVE:
            edge = self.root.vertex.get_edge_to(node.vertex)
            if edge and edge.get_state() == State.ACTIVE:
                edge.set_state(State.NONE)
        # faz o caminho de volta para a raiz e ativa os vértices e arestas   
        n = node
        while n.parent:
            if n.vertex.get_state() != State.ACTIVE:
                self.new_vertices += 1
            n.vertex.set_state(State.ACTIVE)
            n.edge.set_state(State.ACTIVE)
            n = n.parent

    def destroy(self, node):
        if not node.children:
            return
        for child in node.children:
            if child:
                self.destroy(child)
                child.destroy()
        node.children = None


# NEMERTEA --------------------------------------------------------------------------

def nemertea():
    app.log("#Nemertea starting...")
    prev = None  # Vértice anterior ao atual
    path_size = 1  # Número de vértices no caminho
    vertex_size = app.get_vertex_size()  # Número de vértices no grafo

    # Seleciona um vértice aleatório para começar
    first_vertex = app.get_vertex(random.randint(0, vertex_size-1))
    current = first_vertex
    start_time = time.time()
    current.set_state(State.ACTIVE)  # Define o estado do vértice atual como ativo

    # Se é caminho Hamiltoniano (não ciclo), cria o primeiro caminho a partir do vértice inicial
    if not app.get_var("cycle"):
        path_size += first_path(current)

    first = True
    while True:
        while True:
            size = NBFS(current, first).run()
            path_size += size

            if app.is_stopped():
                return 0

            if size == 0:
                break
            first = False

        next = next_vertex(prev, current)
        prev = current
        current = next

        if app.is_stopped():
            return 0

        if current is None or current.get_id() == first_vertex.get_id() or path_size == vertex_size:
            break

    end_time = time.time()
    execution_time = end_time - start_time
    app.set_execution_time(execution_time)
    app.log(f"#Execution time: {execution_time} seconds")
    return path_size


def next_vertex(prev, current):
    edge_size = current.get_edge_size()
    for i in range(edge_size):
        e = current.get_edge(i)
        if e.get_state() == State.ACTIVE:
            adjacent = current.get_adjacent(e)
            if prev is None or adjacent.get_id() != prev.get_id():
                return adjacent
    return None


def first_path(root):
    root.set_state(State.ACTIVE)
    edge_size = root.get_edge_size()
    for i in range(edge_size):
        app.step()
        edge = root.get_edge(i)
        if edge.get_state() != State.ACTIVE:
            adjacent = root.get_adjacent(edge)
            if adjacent.get_state() != State.ACTIVE:
                edge.set_state(State.ACTIVE)
                return first_path(adjacent) + 1
        if app.is_stopped():
            return 0
    return 0


def evaluate(path_size):
    vertex_size = app.get_vertex_size()
    remaining_vertices = vertex_size - path_size
    if vertex_size == path_size:
        app.set_solved(True)
        app.log("#The algorithm nemertea solves the Hamiltonian path problem for this graph.")
    else:
        app.set_solved(False)
        app.log("#The algorithm nemertea did not solve the Hamiltonian path problem for this graph.")
        app.log(f"#Vertex on path: {path_size} of {vertex_size}, {remaining_vertices} left.")


# Inicialização do algoritmo

random.seed(time.time())
path_size = nemertea()
evaluate(path_size)

File: scripts/nemertea/nemertea.lua
dofile("scripts/nemertea/nbfs.lua")

--- Nemertea Algorithm
---@author Saulo Popov Zambiasi
---@language Lua
---@version 1.2
---@since 10/10/2021
---@update 08/05/2025
---@description
---
--- Nemertea is an algorithm for solving the Hamiltonian cycle
--- problem in graphs. It is a heuristic algorithm that
--- uses territorial conquest as a strategy. The algorithm
--- starts in a random closed region and adds new closed regions
--- until all vertices are part of the boundaries of that
--- territory. To add new regions, a vertex v is taken from the
--- frontier and follows paths and external vertices of the
--- frontier until it finds a vertex u that is a neighbor of v
--- and is on the frontier, with no other vertices between them.
--- The path taken from v to u through the external area is
--- added as a new frontier and the boundary that was between v
--- and u is undone. In this way, a new area is added to the
--- territory. To find this new region, the algorithm uses a
--- custom Breadth First Search.
---
--- This lua code is executed in the Grafuria software for
--- creating and editing graphs and executing algorithms on
--- the graphs:
---
--- https://github.com/saulopz/grafuria
---

--- Executes the Nemertea algorithm.
--- @return number Returns the number of vertices in the path.
function nemertea()
    app:log("#Nemertea starting...")
    local prev = nil -- Vertex before the current one
    local path_size = 1 -- Number of vertices in the path
    local vertex_size = app:get_vertex_size() -- Number of vertices in the graph
    -- Randomly selects a vertex to start the algorithm
    local first_vertex = app:get_vertex(math.random(1, vertex_size))
    local current = first_vertex -- Starting current vertex at the first
    local start_time = os.clock()
    current:set_state(State.ACTIVE) -- Sets the current status to active
    local first = true

    -- If is a Hamiltonian Path, we create first path from the
    -- initial vertex until not find a valid way. If it is a
    -- Hamiltonian Cycle, we do not execute this
    if not app:get_var("cycle") then
        -- Adds the first path to the current vertex
        path_size = path_size + first_path(first_vertex)
        first = false
    end

    repeat -- Scans active vertices to conquer new regions    
        -- Run NBFS to expand the PATH until no new vertices are added
        repeat
            -- Run NBFS to expand the PATH until no new vertices are added
            local size = NBFS.new(current, first):run()
            path_size = path_size + size

            -- If command stops, exit the algorithm
            if app:is_stopped() then
                return 0
            end
            first = false
        until size == 0 -- NBFS not create a new path

        -- Go to the next vertex without returning to prev.
        local next = next_vertex(prev, current)
        prev = current
        current = next

        -- If command stops, exit the script
        if app:is_stopped() then
            return 0
        end

        -- It returned to the initial vertex or the path
        -- size is equal to the number of vertices in the graph.
    until not current or current:get_id() == first_vertex:get_id() or path_size == vertex_size
    local end_time = os.clock()
    local execution_time = end_time - start_time
    app:set_execution_time(execution_time)
    app:log("#Execution time: " .. execution_time .. " seconds")
    return path_size
end

--- Returns the next vertex in the graph that is on the frontier.
--- prev is passed as a parameter so that the algorithm does not
--- follow where it came from, which can cause an unwanted loop.
---@param prev Vertex previous vertex
---@param current Vertex current vertex
---@return Vertex Returns the next vertex in the path.
function next_vertex(prev, current)
    local edge_size = current:get_edge_size()
    for i = 1, edge_size do
        local e = current:get_edge(i)

        -- If edge is Active, then it's in PATH
        if e:get_state() == State.ACTIVE then
            local adjacent = current:get_adjacent(e)

            -- If prev is nil, then we are in the first iteration or
            -- if the adjacent vertex is not equal to the previous one,
            -- then we can return it.
            if prev == nil or adjacent:get_id() ~= prev:get_id() then
                return adjacent
            end
        end
    end
    return nil
end

--- First path of the graph.
--- This function is used to find the first path of the graph
--- in case of Hamiltonian path. Hamiltonian cycle not use this
--- @param root Vertex root vertex
--- @return number Returns the number of vertices in the path.
function first_path(root)
    local size = 0
    local p = root
    repeat
        app:step()
        local found = false
        local edge_size = p:get_edge_size()
        local i = 1
        while (i <= edge_size and not found) do
            local edge = p:get_edge(i)
            if edge:get_state() ~= State.ACTIVE then
                local adjacent = p:get_adjacent(edge)
                if adjacent:get_state() ~= State.ACTIVE then
                    edge.set_state(State.ACTIVE)
                    adjacent.set_state(State.ACTIVE)
                    p = adjacent
                    found = true
                    size = size + 1
                end
            end
            i = i + 1
        end
    until found == false
    return size
end

--- Evaluates whether the algorithm was successful.
--- If the algorithm reaches all vertices, it has
--- found a Hamiltonian cycle in the graph. Otherwise,
--- it is undefined.
--- @param path_size number Number of vertices in the path
function evaluate(path_size)
    local vertex_size = app:get_vertex_size()
    local remaining_vertices = vertex_size - path_size
    if vertex_size == path_size then
        app:set_solved(true)
        app:log("#The algorithm nemertea solves the Hamiltonian path problem for this graph.")
    else
        app:set_solved(false)
        app:log("#The algorithm nemertea not to solve the Hamiltonian path problem for this graph.")
        app:log("#Vertex on path: " .. path_size .. " of " .. vertex_size .. ", " .. remaining_vertices .. " left.")
    end
end

-- =========================
-- Nemertea initialization
-- =========================

-- Initializes the random number generator seed
math.randomseed(os.time())
path_size = nemertea()
evaluate(path_size)
